modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosInstrumentActions.scala

```diff
@@ -78,130 +96,128 @@ class GmosInstrumentActions[F[_]: MonadError[?[_], Throwable]: Logger, A <: Gmos
+          .flatMap(inst.continueCommand)
+      )
+      .as(Result.Partial(NSContinue))
+
+  // Calculate the subexposures
+  private def subexposures(
+    cycles: Int
+  ): List[Subexposure] =
+    (for {
+      i     <- 0 until cycles
+      j     <- 0 until Gmos.NsSequence.length
+      stage = Gmos.NsSequence.toList.lift(j).getOrElse(StageA)
+    } yield Subexposure(cycles, i, j, stage)).toList
 
   /**
-   * Calculate the actions for a N&S cycle
-   */
-  private def actionPositions(
-    env:   ObserveEnvironment[F],
-    rows:  Int,
+    * Stream of actions of one sub exposure
+    */
+  def oneSubExposure(
+    fileId:    ImageFileId,
+    sub:       Subexposure,
+    rows:      Int,
     positions: Vector[NSPosition],
-    post:  (Stream[F, Result[F]], ObserveEnvironment[F]) => Stream[F, Result[F]]
-  ): List[ParallelActions[F]] =
-    Gmos.NsSequence.zipWithIndex
-      .map {
-        case (stage, i) =>
-          val rowsToShuffle = if (stage === StageA) 0 else rows
-          val nsPositionO = positions.find(_.stage === stage)
-          List(
-            // Configure rows to shuffle
-            NonEmptyList(
-              inst
-                .configureShuffle(rowsToShuffle)
-                .as(Response.Configured(inst.resource))
-                .toAction(ActionType.Configure(inst.resource)),
-              (env.getTcs, nsPositionO).mapN{ case (tcs, nsPos) =>
-                tcs.nod(
-                  stage,
-                  InstrumentOffset(
-                    tag[OffsetP](nsPos.offset.p.toRadians.radians), tag[OffsetQ](nsPos.offset.q.toRadians.radians)),
-                  nsPos.guide === Guiding.Guide
-                ).toAction(ActionType.Configure(tcs.resource))
-              }.toList
-            ),
-            // Do an obs observe/continue
-            NonEmptyList.one(
-              Action(
-                ActionType.Observe,
-                post( // post lets upstream to mix progress events
-                  safeObserve(i, env),
-                  env
+    env:       ObserveEnvironment[F],
+    post:      (Stream[F, Result[F]], ObserveEnvironment[F]) => Stream[F, Result[F]]
+  ): Stream[F, Result[F]] = {
+    val rowsToShuffle = if (sub.stage === StageA) 0 else rows
+    val nsPositionO   = positions.find(_.stage === sub.stage)
+    // Configure GMOS rows
+    Stream.eval(
+      inst.configureShuffle(rowsToShuffle).as(Result.Partial(NSRowsConfigure))
+    ) ++
+      // TCS Nod
+      (env.getTcs, nsPositionO).mapN {
+        case (tcs, nsPos) =>
+          Stream.eval(
+            tcs
+              .nod(
+                sub.stage,
+                InstrumentOffset(
+                  tag[OffsetP](nsPos.offset.p.toRadians.radians),
+                  tag[OffsetQ](nsPos.offset.q.toRadians.radians)
                 ),
-                Action.State(Action.ActionState.Idle, Nil)
+                nsPos.guide === Guiding.Guide
               )
-            )
+              .as(Result.Partial(NSTCSNod))
```

Comment:
Shouldn't the result have the stage?

Reply:
I want to add more data once the UI is being developed. but yes we should add more info about where we are